Index the wallet's UTxO set by asset type

jonathanknowles opened the pull request:
# Issue Number

ADP-633

# Overview

This PR adds a `UTxOIndex` type, which:

* Provides a mapping from assets to sets of transaction inputs. A transaction input appears in the set if and only if its corresponding output contains a non-zero quantity of that asset.

* Makes it possible to efficiently compute the subset of a UTxO set containing a particular asset, or to select a UTxO entry containing that asset, without having to search linearly through the entire UTxO set.

* Keeps track of the current UTxO balance for all assets, making it possible to efficiently look up the quantity of a particular asset without having to sum across the entire UTxO set.

# Motivation

The wallet currently uses the **Random-Improve** algorithm to generate the majority of its coin selections. This algorithm works by choosing entries at random from the wallet's entire UTxO set until there are enough entries to cover the ada value of each desired output, and then attempts to improve upon each selection until its total value is around twice the value of each output (so as to generate useful change outputs that match the wallet owner's typical spending pattern).

However, in a multi-asset context where we're attempting to cover some amount of a given token, it's imperative that we have a way to efficiently search a subset of a UTxO for entries containing a particular asset without having to traverse the entire set.

# Design

## Indexing Modes

The `UTxOIndex` type supports two possible indexing modes:

1. **Incremental indexing**: entries are added over time via repeated applications of the `insert` operation.

1. **Just-in-time indexing**: the index is constructed from an existing UTxO set in one pass, using the `fromUTxO` operation.

We expect the first version of the MA coin selection algorithm to make use of _just-in-time_ indexing.

## Operations

### Construction

```hs
fromUTxO
    :: UTxO -> UTxOIndex
```

### Insertion

```hs
insert
    :: TxIn -> TxOut -> UTxOIndex -> UTxOIndex
insertMany
    :: Foldable f => f (TxIn, TxOut) -> UTxOIndex -> UTxOIndex
```

### Deletion

```hs
delete
    :: TxIn -> UTxOIndex -> UTxOIndex
deleteMany
    :: Foldable f => f TxIn -> UTxOIndex -> UTxOIndex
```

### Random Selection

```hs
-- | Selects an entry at random from the index according to the given filter.
--
-- Returns the selected entry and an updated index with the entry removed.
--
-- Returns 'Nothing' if there were no matching entries.
--
selectRandom
    :: MonadRandom m
    => UTxOIndex
    -> SelectionFilter
    -> m (Maybe ((TxIn, TxOut), UTxOIndex))

-- | Specifies a filter for selecting UTxO entries.
--
data SelectionFilter
    = Any
        -- ^ Select any UTxO entry from the entire set.
    | WithAsset AssetId
        -- ^ Select any UTxO entry that has a non-zero quantity of the specified
        -- asset.
```

jonathanknowles committed:
Add generators for small, strictly-positive token quantities.

jonathanknowles committed:
Add laws for the `PartialOrd` class.

jonathanknowles committed:
Add `PartialOrd` instance for `TokenBundle`.

jonathanknowles committed:
Add `Read` instance for `NonEmptyMap`.

jonathanknowles committed:
Add `Read` instances for several wallet primitive types.

- `Address`
- `AssetId`
- `Coin`
- `TokenBundle`
- `TokenMap`
- `TxIn`
- `TxOut`

jonathanknowles committed:
Add strict non-empty set type.

jonathanknowles committed:
Add indexed UTxO type.

The index makes it possible to efficiently compute the subset of a UTxO
set containing a particular asset, or to select a UTxO entry containing
that asset, without having to search through the entire UTxO set.

The index also keeps track of the current UTxO balance for all assets,
making it possible to efficiently look up the quantity of a particular
asset without having to sum across the entire UTxO set.

jonathanknowles committed:
Add function `subtract` to `TokenQuantity`.

jonathanknowles committed:
Add function `subtract` to `TokenMap`.

jonathanknowles committed:
Add function `subtract` to `TokenBundle`.

jonathanknowles committed:
Add function `getAssets` to `TokenBundle`.

jonathanknowles committed:
Add `Monoid` instance for `TokenBundle`.

jonathanknowles committed:
Add generator module for `Address` type.

jonathanknowles committed:
Add generator module for transaction-related types.

jonathanknowles committed:
Add function `getAssets` to `TokenMap`.

jonathanknowles committed:
Add `PartialOrd` instance for `TokenMap`.

jonathanknowles committed:
Add `NFData` instance for `AssetId`.

jonathanknowles committed:
Fork `subtract` functions into safe and unsafe variants.

This change explicitly forks the various `subtract` functions into safe and
unsafe variants for the following types:

- `TokenQuantity`
- `TokenMap`
- `TokenBundle`

The safe variants all adhere to the following pattern:

```
-- Returns 'Nothing' if the pre-condition is violated.
subtract :: T -> T -> Maybe T
```

The unsafe variants all adhere to the following pattern:

```
-- Throws a run-time exception if the pre-condition is violated.
unsafeSubtract :: T -> T -> T
```

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+--
+-- See the documentation for 'UTxOIndex' for more details.
+--
+module Cardano.Wallet.Primitive.Types.UTxOIndex.Internal
```
Internal modules should expose everything. They're already labelled `Internal`, so there's no reason to e.g. not export the constructor (e.g. for testing purposes or whatnot).

jonathanknowles replied:
Just for the record: in a review call we agreed that leaving it as is would not be harmful.

If there is a need for someone to access the internals of this data type, it would be easy to create a patch that exports the default data constructor.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex.hs:
```diff
@@ -0,0 +1,51 @@
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Provides the 'UTxOIndex' type, which stores a UTxO set indexed by asset ID.
```
I find this a little hard to understand if I assume I didn't know anything about it yet.

jonathanknowles replied:
Hi Julian. Thanks for this feedback.

I've tried to make this a bit clearer in commit 51f535da97bf9838e1e7397a3092a974aaeb763d.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex.hs:
```diff
@@ -0,0 +1,51 @@
+-- |
+-- Copyright: © 2018-2020 IOHK
```
2021

jonathanknowles replied:
Fixed in 0ef36c8a6243c24fdd5ebcfce258c416748ddfab.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+{-# LANGUAGE OverloadedLabels #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
```
2021

jonathanknowles replied:
Fixed in 0ef36c8a6243c24fdd5ebcfce258c416748ddfab.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+-- A transaction input appears in the set if and only if its corresponding
+-- output contains a non-zero quantity of that asset.
+--
+-- The index makes it possible to efficiently compute the subset of a UTxO
```
Right, this is the motivation. It's more important then the rest of the comments. We should highlight this more.
Maybe also add that this is used for coin selection, which is the reason for all this.

jonathanknowles replied:
Hi Julian. Thanks for this feedback!

I've tried to make this a bit clearer in commit 51f535da97bf9838e1e7397a3092a974aaeb763d.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+    , balance
+        :: !TokenBundle
+    , utxo
+        :: !(Map TxIn TxOut)
```
Why is this not using `newtype UTxO = UTxO { getUTxO :: Map TxIn TxOut }`?

jonathanknowles replied:
> Why is this not using newtype UTxO = UTxO { getUTxO :: Map TxIn TxOut }?

We could, but it would add a layer of wrapping and unwrapping to every operation that manipulates the `utxo` field (which is internal).

The current implementation only needs to do this wrapping and unwrapping just once, if we actually need to convert to from an ordinary `UTxO`:
```
toUTxO :: UTxOIndex -> UTxO
toUTxO = UTxO . utxo

fromUTxO :: UTxO -> UTxOIndex
fromUTxO = Map.foldlWithKey' (\u i o -> insertUnsafe i o u) empty . getUTxO
```

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+--
+empty :: UTxOIndex
+empty = UTxOIndex
+    { index = mempty
```
I'm generally not a fan of `mempty`. Using type-specific values is more clear, imo. I mean, there's a reason you provided `empty` and not just a Monoid instance.

jonathanknowles replied:
Fixed in 00770944363b5c0fd2b9af5cd210a8c03f0e1a43.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+    }
+
+-- | Creates a singleton index from the specified input and output.
+--
```
Why all these semi-empty lines?

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs:
```diff
@@ -0,0 +1,488 @@
+
+-- | The result of checking the invariant with the 'checkInvariant' function.
+--
+data InvariantStatus
```
I'd suggest to also explain the invariants right here in the data types. Then it's an easy overview of all of them, rather then searching for the functions.

jonathanknowles replied:
Hopefully this is fixed by 368015f7af854d9af23bc8a07250a7cf4e615d09.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs:
```diff
@@ -157,6 +166,11 @@ instance TypeError ('Text "Ord not supported for token maps")
         => Ord TokenMap where
     compare = error "Ord not supported for token maps"
 
+instance PartialOrd TokenMap where
```
We should explain why we need this `PartialOrd` and a simplistic example why `Ord` doesn't work.

jonathanknowles replied:
> We should explain why we need this `PartialOrd` and a simplistic example why `Ord` doesn't work.

Fixed in 07116a8d489ff73d751c295089148df70ae8783b.

hasufell reviewed (commented):
This is a partial review. Still have to look at the tests in more detail. But so far it looks good to me.

hasufell commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs:
```diff
@@ -0,0 +1,531 @@
+prop_delete_invariant i u = invariantHolds $ UTxOIndex.delete i u
+
+prop_selectRandom_invariant :: UTxOIndex -> SelectionFilter -> Property
+prop_selectRandom_invariant i f = monadicIO $ do
```
Can we try avoid IO and use ` DRG gen => MonadRandom (MonadPseudoRandom gen)` instead? I also see this in many other places.

jonathanknowles replied:
> Can we try avoid IO and use DRG gen => MonadRandom (MonadPseudoRandom gen) instead? I also see this in many other places.

I agree that it would be nicer to avoid `IO` here.

Just for the record: we held a review call and agreed that it would be acceptable to refactor this in a future PR.

hasufell commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs:
```diff
@@ -0,0 +1,531 @@
+
+prop_delete_balance :: TxIn -> UTxOIndex -> Property
+prop_delete_balance i u =
+    UTxOIndex.balance (UTxOIndex.delete i u) === expected
```
What's the chance `TxIn` appears in `UTxOIndex` here? I'm not sure we have a reasonable distribution.

We should probably use [cover](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:cover) here.

jonathanknowles replied:
> What's the chance TxIn appears in UTxOIndex here?

I've added coverage checks in e99202270688d298d0a5d07f5dfb1b9dadb9bb94.

hasufell commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs:
```diff
@@ -0,0 +1,531 @@
+    expected = case selectionFilter of
+        Any ->
+            Just ((i, o), UTxOIndex.empty)
+        WithAsset a | txOutHasAsset o a ->
```
What's the distribution of this?

jonathanknowles replied:
> What's the distribution of this?

It's around 1:1:

```
    Index Selection
      prop_SelectionFilter_coverage
        +++ OK, passed 200 tests:
        50.5% selectionFilter: WithAsset
        49.5% selectionFilter: Any
```

I've added a coverage check in ad53a80907287f00710a5816ee00dafe48494030.

hasufell commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs:
```diff
@@ -0,0 +1,531 @@
+-- Returns a list of all the entries that matched, and an updated index with
+-- the selected entries removed.
+--
+selectAll
```
I'm seriously wondering why this is not in `Cardano.Wallet.Primitive.Types.UTxOIndex.Internal`.

jonathanknowles replied:
> I'm seriously wondering why this is not in
> Cardano.Wallet.Primitive.Types.UTxOIndex.Internal.

There's really no use for it outside of tests.

hasufell commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs:
```diff
@@ -0,0 +1,531 @@
+--
+-- This should always succeed, provided the index is not empty.
+--
+prop_selectRandom_one_any :: UTxOIndex -> Property
```
Alright, this is the property test that actually tests that the implementation of the function doesn't misbehave (e.g. falsely returning Nothing).

But, it relies on other functions not being buggy (such as `delete` etc). There's a very low chance a bug may slip through. This is why I think we must have unit tests as well. Property tests are not a substitute for unit tests.

hasufell commented on lib/strict-non-empty-containers/src/Data/Set/Strict/NonEmptySet.hs:
```diff
@@ -0,0 +1,95 @@
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Provides a strict implementation of a non-empty set.
```
What was the motivation we added this module?

jonathanknowles replied:
> What was the motivation we added this module?

It's used here:

```hs
data UTxOIndex = UTxOIndex
    { index
        :: !(Map AssetId (NonEmptySet TxIn))
    , balance
        :: !TokenBundle
    , utxo
        :: !(Map TxIn TxOut)
    }
    deriving (Eq, Generic, Read, Show)
```

Using a non-empty set has two advantages:

1. We don't have to manually enforce the invariant that `(asset, [])` is not allowed in the index.
2. The `Eq` instance can be auto-derived.

hasufell reviewed (commented):
Reviewed the tests. I think it looks pretty good, but I'd add a few unit tests to test the implementation of e.g. `selectRandom`. I explained my reasoning in one of [the comments](https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r554739785).

jonathanknowles committed:
Export `empty` from type `TokenBundle`.

jonathanknowles committed:
Use type-specific `empty` functions in definition of `UTxOIndex.empty`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553771063

jonathanknowles committed:
Change copyright upper bounds to 2021.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553767922
https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553767942

jonathanknowles committed:
Add explanatory comment to `PartialOrd` instance for `TokenMap`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553789000

jonathanknowles committed:
Add coverage checks to tests of `delete` and `insert` operations.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r554695305

jonathanknowles committed:
Add coverage check for `SelectionFilter` type.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r554708244

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs:
```diff
@@ -0,0 +1,573 @@
+        assert $ selected `Set.member` set
+  where
+    confidence = stdConfidence {tolerance = 0.99}
+    elementCount = greatest - smallest
```
This is an off-by-one error.

jonathanknowles replied:
> This is an off-by-one error.

Fixed in 370ae27087ef15548e38b0cd7b5416593cc4db41.

jonathanknowles committed:
Fix off-by-one error in `prop_selectRandomSetMember_coversRangeUniformly`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431/files#r554889471

jonathanknowles committed:
Add motivation to the module header comment for type `UTxOIndex`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553767667
https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553769984

jonathanknowles committed:
Add documentation to constructors of `InvariantStatus`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2431#discussion_r553784610

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Gen.hs:
```diff
@@ -0,0 +1,109 @@
+genTxHashSmallRange = elements txHashes
+
+shrinkTxHashSmallRange :: Hash "Tx" -> [Hash "Tx"]
+shrinkTxHashSmallRange hash = filter (< hash) txHashes
```
Small remark: I am not exactly sure what this is shrinking. Unless I am missing something, there's no way to shrink a hash.

jonathanknowles replied:
> Small remark: I am not exactly sure what this is shrinking. Unless I am missing something, there's no way to shrink a hash.

In the case of `genTxHashSmallRange`, hashes are initially chosen from a small fixed range, so as to allow collisions:

```hs
txHashes :: [Hash "Tx"]
txHashes = mkTxHash <$> ['0' .. '7']
```

The shrinker doesn't shrink the hashes themselves: it merely returns a small list of hashes that are _less than_ the given hash in the available set.

```hs
> generate genTxHashSmallRange
> a = it
> toText a
  "4444444444444444444444444444444444444444444444444444444444444444"

> toText <$> shrinkTxHashSmallRange a
[ "0000000000000000000000000000000000000000000000000000000000000000"
, "1111111111111111111111111111111111111111111111111111111111111111"
, "2222222222222222222222222222222222222222222222222222222222222222"
, "3333333333333333333333333333333333333333333333333333333333333333"
]
```

In some cases, this can lead to simpler counterexamples.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex.hs:
```diff
@@ -0,0 +1,55 @@
+--
+-- The index makes it possible to efficiently compute the subset of a UTxO
+-- set containing a particular asset, or to select a UTxO entry containing
+-- that asset, without having to search through the entire UTxO set.
```
I have some reservations about this entire module. The claim seems to be that computing a subset of a UTxO containing a particular asset id would be too slow, yet.. is that really the case?

Partitioning a UTxO can be done in one pass, and is only needed before running a coin-selection. Even for large UTxO sets this should already be pretty fast to do, and it's a cost we would pay only for coin selections. Now, we'd be paying the cost of the UTxOIndex for every single block application. Although it may be small, it's also impacting a much larger part of the code for something that is only truly needed at one (infrequent) point in time 

Wouldn't it be easier to "flatten" the `TokenMap` instead to facilitate the partitioning, and do the partitioning when setting up the context for coin selections:

```hs
newtype TokenMap = TokenMap
    { unTokenMap
        :: Map AssetId TokenQuantity
    }
```

```hs
partitionUTxO :: UTxO -> NonEmpty AssetId -> Map AssetId UTxO
```

jonathanknowles replied:
@KtorZ wrote:
> I have some reservations about this entire module. The claim seems to be that computing a subset of a UTxO containing a particular asset id would be too slow, yet.. is that really the case?
> Partitioning a UTxO can be done in one pass, and is only needed before running a coin-selection. Even for large UTxO sets this should already be pretty fast to do, and it's a cost we would pay only for coin selections. Now, we'd be paying the cost of the UTxOIndex for every single block application. Although it may be small, it's also impacting a much larger part of the code for something that is only truly needed at one (infrequent) point in time

It seems we're comparing:

- **incremental indexing**: where we update the index in small chunks, as we parse blocks.
- **just-in-time indexing**: where we build the entire index in one go, before each coin selection.

If we believe that "just-in-time indexing" is acceptable for now, then we could simply remove the following commit from this PR, which integrates `UTxOIndex` into the `Wallet` type:

https://github.com/input-output-hk/cardano-wallet/pull/2431/commits/43bbeecd580dd0d6952223cbc815ec6c380884e6

This of course leaves us with the option of re-integrating it later on, if we discover (through benchmarking) that incremental indexing is preferable (perhaps for users of wallets with very large UTxO sets).

> Wouldn't it be easier to "flatten" the TokenMap instead to facilitate the partitioning, and do the partitioning when setting up the context for coin selections:
> ```
> newtype TokenMap = TokenMap
>    { unTokenMap
>        :: Map AssetId TokenQuantity
>    }
> partitionUTxO :: UTxO -> NonEmpty AssetId -> Map AssetId UTxO
> ```

This is exactly the type of feedback that would have been useful at the design stage.

Unfortunately, we're now very close to the deadline, and I suspect there is no time budget for making fundamental alterations to the design.

There are course many different ways to slice this problem.

When selecting a UTxO entry at random that contains a given asset, if we choose to select that entry, then we also need to remove it from consideration by future selections. In your proposed simplification, we'd need to remove the entry from all the values of the map returned by `partitionUTxO`.

But this is essentially what the `UTxOIndex` type (and `selectRandom` operation) already does. In addition, it also keeps track of the current balance, which is something we need to keep track of when running coin selection (both for the selected and remaining UTxO sets).

Given that we're under a tight deadline, and given that the existing coin selection code is built with the `UTxOIndex` type in mind, and given that it is fully tested and already been through the review process, I propose that we:

1. Keep the existing implementation of `UTxOIndex`.
2. Remove the commit integrating it into the `Wallet` type.
3. Create the index only when we need to run a coin selection.

Thoughts?

hasufell replied:
> Thoughts?

Yes
